util/etcd/etcd.go

```diff
@@ -0,0 +1,368 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/etcdserver/etcdserverpb"
+	"google.golang.org/grpc"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const etcdTimeout = 2 * time.Second
+
+// exponential backoff for etcd operations, values taken from kubeadm
+var etcdBackoff = wait.Backoff{
+	Steps:    9,
+	Duration: 50 * time.Millisecond,
+	Factor:   2.0,
+	Jitter:   0.1,
+}
+
+type grpcDial = func(ctx context.Context, addr string) (net.Conn, error)
+
+// Client provides connection parameters for an etcd cluster
+type Client struct {
+	log           logr.Logger
+	timeout       time.Duration
+	backoffParams wait.Backoff
+	endpoint      string
+	etcdClient    *clientv3.Client
+}
+
+// MemberAlarm represents an alarm type association with a cluster member
+type MemberAlarm struct {
+	// MemberID is the ID of the member associated with the raised alarm
+	MemberID uint64
+
+	// Type is the type of alarm which has been raised.
+	Type AlarmType
+}
+
+type AlarmType int32
+
+const (
+	// AlarmOK denotes that the cluster member is OK
+	AlarmOk AlarmType = iota
+	// AlarmNoSpace denotes that the cluster member has run out of disk space
+	AlarmNoSpace AlarmType = iota
+	// AlarmCorrupt denotes that the cluster member has corrupted data
+	AlarmCorrupt AlarmType = iota
+)
+
+// Adapted from kubeadm
+
+// Member struct defines an etcd member; it is used to avoid spreading
+// github.com/coreos/etcd dependencies
+type Member struct {
+	// ID is the ID of this cluster member
+	ID uint64
+	// Name is the human-readable name of the member. If the member is not started, the name will be an empty string.
+	Name string
+	// PeerURLs is the list of URLs the member exposes to the cluster for communication.
+	PeerURLs []string
+	// ClientURLs is the list of URLs the member exposes to clients for communication. If the member is not started, clientURLs will be empty.
+	ClientURLs []string
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool
+	// Alarms is the list of alarms for a member
+	Alarms []AlarmType
+}
+
+type Status struct {
+	// Leader is the member ID which the responding member believes is the current leader.
+	Leader uint64
+
+	// Version is the cluster protocol version used by the responding member
```

Comment:
```suggestion
	// Version is the cluster protocol version used by the responding member.
```

Consistency with the other comments and the rest of the repository.